[fix] 修复复数log二阶导数计算错误

scyyh11 opened the pull request:
<!-- TemplateReference: https://github.com/PaddlePaddle/Paddle/wiki/PULL-REQUEST-TEMPLATE--REFERENCE -->
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->

### PR Category
<!-- One of [ User Experience | Execute Infrastructure | Operator Mechanism | CINN | Custom Device | Performance Optimization | Distributed Strategy | Parameter Server | Communication Library | Auto Parallel | Inference | Environment Adaptation ] -->
Operator Mechanism

### PR Types
<!-- One of [ New features | Bug fixes | Improvements | Performance | BC Breaking | Deprecations | Docs | Devs | Not User Facing | Security | Others ] -->
Bug fixes

### Description
<!-- Describe what you've done -->

[used AI Studio]

**修复复数log高阶导数计算错误**

**问题描述:**
- 复数log的二阶导数测试失败，实际结果与期望值虚部符号相反
- 测试期望: `ddx = conj(-1/x^2)`，但实际得到: `ddx = -1/x^2` (缺少共轭)

**根本原因:**
- 复合实现(composite)使用实值公式: `ddout = ddx/x`, `dx = -dout·ddx/x^2`
- 忽略了复数数学中的共轭运算，导致分母缺少`conj(x)`
- 内核实现已正确处理复数共轭，但调度系统优先选择复合路径
- 复合路径无法调用`conj`算子，因为未在prim api.yaml中注册

**技术分析:**
```yaml
# 问题：composite无法表达复数共轭运算
- backward_op : log_double_grad
  kernel:
    func : log_double_grad
  composite : log_double_grad(x, grad_out, grad_x_grad, x_grad, grad_out_grad)
```

**解决方案:**
1. **添加conj到prim算子集合**: 在`paddle/fluid/prim/api/api.yaml`中添加`conj`算子
2. **修复composite实现**: 更新`composite_backward_api.h`和`composite_double_backward_api.h`中的log梯度函数

**代码修改:**
```diff
# 1. 添加conj到prim api.yaml
+ - conj

# 2. 修复composite log_grad
- set_output<T>(out_grad / x, x_grad);
+ set_output<T>(out_grad / conj<T>(x), x_grad);

# 3. 修复composite log_double_grad  
- auto x_grad_tmp = -grad_out / (x * x) * grad_x_grad;
- auto grad_out_grad_tmp = grad_x_grad / x;
+ auto conj_x = conj<T>(x);
+ auto x_grad_tmp = -(grad_out * grad_x_grad) / (conj_x * conj_x);
+ auto grad_out_grad_tmp = grad_x_grad / conj_x;
```

**数学原理:**
- **复数log一阶导数**: `d/dx log(x) = 1/conj(x)`
- **复数log二阶导数**: `d²/dx² log(x) = -1/conj(x)²`, `d²/dxdy log(x) = 1/conj(x)`
- **实数情况**: `conj(x) = x`，公式退化为标准实值公式

**相关Issue：**
[#75393](https://github.com/PaddlePaddle/Paddle/issues/75393)

scyyh11 committed:
fix: remove composite implementations for log_grad and log_double_grad

- Remove composite entries from backward.yaml to use correct kernel implementations.

scyyh11 committed:
Merge branch 'PaddlePaddle:develop' into fix/LogGradGradFunctor

YqGe585 reviewed (approved):
LGTM

HydrogenSulfate commented on paddle/phi/ops/yaml/backward.yaml:
```diff
@@ -2046,7 +2046,6 @@
     param : [x, x]
   kernel :
     func : log_double_grad
-  composite : log_double_grad(x, grad_out, grad_x_grad, x_grad, grad_out_grad)
```
这里不能直接删除，否则log算子的三阶及以上反向就无法使用了

HydrogenSulfate commented on paddle/phi/ops/yaml/backward.yaml:
```diff
@@ -2060,7 +2059,6 @@
   kernel :
     func : log_grad
   backward : log_double_grad
-  composite : log_grad(x, out_grad, x_grad)
```
这个可以保留

HydrogenSulfate reviewed (commented):
比较好的解决办法应该是把`conj`算子加到`api.yaml`(动态图基础算子集合里)，然后给`composite_double_backward_api.h`和`composite_backward.h`下的`log_grad`和`log_double_grad`两个算子内正确的位置加入`conj`运算。这样删除其实没有解决根本问题

scyyh11 committed:
fix: fix complex log higher-order grads via conj in composite

- Add conj to prim api.yaml
- Use conj(x) in composite log_grad and log_double_grad
- Re-enable composite entries in backward.yaml to keep higher-order grads working

scyyh11 commented:
/re-run all-failed

scyyh11 committed:
reset: Reset backward.yaml to commit 676eb2a6c1d8bc19dffed1b8aae7ac1f58b17a5e

HydrogenSulfate commented:
> /re-run all-failed

能补充一下单测吗，至少在PR里贴一下复数+log一阶、二阶、三阶微分跟torch对比的结果

scyyh11 commented:
### 补充

```python
import numpy as np
import paddle

np.random.seed(2024)
shape = (5, 7)

# Complex input in the first quadrant (to avoid branch ambiguities at/near 0)
x_numpy = (np.random.uniform(0.1, 1, shape) + 1j * np.random.uniform(0.1, 1, shape)).astype("complex128")

# Expected analytic forms (with conjugate, matching common complex-grad convention)
expected_dx    = np.conj( 1 / (x_numpy)        )        # d/dx log x = 1/x
expected_ddx   = np.conj(-1 / (x_numpy**2)     )        # d^2/dx^2 log x = -1/x^2
expected_dddx  = np.conj( 2 / (x_numpy**3)     )        # d^3/dx^3 log x =  2/x^3

# Paddle forward & autograd (complex128)
x = paddle.to_tensor(x_numpy, dtype='complex128', stop_gradient=False)
y = paddle.log(x)

# 1st derivative
dx = paddle.grad([y], [x], create_graph=True, retain_graph=True)[0]

# 2nd derivative (keep graph to take third)
ddx = paddle.grad([dx], [x], create_graph=True, retain_graph=True)[0]

# 3rd derivative (final)
dddx = paddle.grad([ddx], [x], retain_graph=True)[0]

dx_np    = dx.numpy()
ddx_np   = ddx.numpy()
dddx_np  = dddx.numpy()

# Quick view
print("NumPy expected dx (top-left 2x2):\n",    expected_dx[:2, :2])
print("Paddle dx (top-left 2x2):\n",            dx_np[:2, :2], "\n")

print("NumPy expected ddx (top-left 2x2):\n",   expected_ddx[:2, :2])
print("Paddle ddx (top-left 2x2):\n",           ddx_np[:2, :2], "\n")

print("NumPy expected dddx (top-left 2x2):\n",  expected_dddx[:2, :2])
print("Paddle dddx (top-left 2x2):\n",          dddx_np[:2, :2], "\n")

# Numeric checks
rtol = 1e-6
atol = 1e-7

def report(name, got, expected):
    diff = np.max(np.abs(got - expected))
    print(f"{name}: max |diff| = {diff:.3e}")

report("dx   ", dx_np,   expected_dx)
report("ddx  ", ddx_np,  expected_ddx)
report("dddx ", dddx_np, expected_dddx)

````

运行结果：

```
NumPy expected dx (top-left 2x2):
 [[1.1967609 +0.68539011j 0.61720269+0.68225656j]
 [0.78038675+0.69848157j 0.52343593+0.84868635j]]
Paddle dx (top-left 2x2):
 [[1.1967609 +0.68539011j 0.61720269+0.68225656j]
 [0.78038675+0.69848157j 0.52343593+0.84868635j]] 

NumPy expected ddx (top-left 2x2):
 [[-0.96247706-1.64049618j  0.08453486-0.84218117j]
 [-0.12112698-1.09017153j  0.44628336-0.88846585j]]
Paddle ddx (top-left 2x2):
 [[-0.96247706-1.64049618j  0.08453486-0.84218117j]
 [-0.12112698-1.09017153j  0.44628336-0.88846585j]] 

NumPy expected dddx (top-left 2x2):
 [[ 0.05495012+5.2459079j  -1.25351755+0.92424403j]
 [-1.33387766+1.87072077j -1.97525917+0.17260071j]]
Paddle dddx (top-left 2x2):
 [[ 0.05495012+5.2459079j  -1.25351755+0.92424403j]
 [-1.33387766+1.87072077j -1.97525917+0.17260071j]] 

dx   : max |diff| = 4.965e-16
ddx  : max |diff| = 1.256e-15
dddx : max |diff| = 1.589e-14
```

scyyh11 commented:
@HydrogenSulfate 已补充，由于pytorch和paddle在同一环境下有冲突，所以我分开补充了

scyyh11 commented:
/re-run all-failed
